### PR TITLE
Backports for 1.5.2

### LINKED ISSUE #37126: ambiguous soft scope warning on anonymous object in version 1.5.0
In Julia 1.5.0, I get a warning about ambiguous scoping of an anonymous object. Following is the shortest code I have found to reproduce this issue:

```julia
julia> include("f.jl")
┌ Warning: Assignment to `#s17` in soft scope is ambiguous because a global variable by the same name exists: `#s17` will be treated as a new local. Disambiguate by using `local #s17` to suppress this warning or `global #s17` to assign to the existing global variable.
└ @ C:\Users\kyungminlee\Code\soft-scope\f.jl:6
```
where `f.jl` is
```julia
function foo()
    f(lhs::Integer, rhs::Integer) = nothing
    f(lhs::Integer, rhs::AbstractVector{<:Integer}) = nothing
    return f
end
struct Bar{T<:Real, P<:Real} end
```

This only seems to happen when I do `include`: no warning is thrown when I put the code in a module and `import` it, or just type it in the repl.

```
julia> versioninfo()
Julia Version 1.5.0
Commit 96786e22cc (2020-08-01 23:44 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)

### LINKED ISSUE #37134: Type declaration on anonymous function arguments in macro not working on 1.5
Hello, 

Originally broke `DistributedArrays`, filed [here](https://github.com/JuliaParallel/DistributedArrays.jl/issues/223). It turns out it is about type declaration of an anonymous function argument inside a macro (that was a mouthful :P).

```
julia> macro bla()
       :(x :: Int -> 62)
       end
@bla (macro with 1 method)

julia> @bla
ERROR: syntax: malformed expression
Stacktrace:
 [1] top-level scope at REPL[2]:1
```

This stopped working on version 1.5, was working as of 1.4 (works without the type declaration `::Int`). Curiously, `eval(:(x::Int -> 62))` works too.

Cheers!

### LINKED ISSUE #37199: One-dimensional views assume their range index is 1-indexed
This happens if the range used to create the view does not span the entire axis of the parent array

```julia
julia> b = zeros(6, 2);

julia> r = Base.IdentityUnitRange(4:6);

julia> b[r, 2] .= 3
3-element view(::Array{Float64,2}, :, 2) with eltype Float64 with indices 4:6:
 3.0
 6.90395899974187e-310
 6.90395899973673e-310

julia> b
6×2 Array{Float64,2}:
 0.0  0.0
 0.0  0.0
 0.0  0.0
 0.0  0.0
 0.0  0.0
 0.0  0.0
```

### LINKED ISSUE #37282: Ambiguity between the representation of Expr containing hexadecimal literals and juxtaposing of a variable starting with 'x'
The two Expr "a" and "b" have two very different meanings but are represented in the same way by the "string" function.
The problem is fortunately only of the representation since both of the Expr get evaluated correctly.
```julia
x01 = 5
#output: 
# 5
a = :((0)x01 + 9)
#output: 
# :(0x01 + 9)
b = :(0x01 + 9)
#output: 
# :(0x01 + 9)
@eval $a
#output: 
# 9
@eval $b
#output: 
# 10

#=
Julia Version 1.5.1
Commit 697e782ab8 (2020-08-25 20:08 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: AMD Ryzen 5 3600 6-Core Processor
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, znver2)
=#
```

### LINKED ISSUE #37401: `Enum` can't be displayed in REPL
trying to show `Enum` in the REPL throws an error, apparently because it's not a `DataType`
seems to be broken from 1.3 to 1.5 (I didn't try on master)

(I stumbled across this when trying to see `subtypes(Any)`

```julia
julia> println(Enum)
Enum

julia> Enum isa DataType
false

julia> Enum
Enum Error showing value of type UnionAll:
ERROR: MethodError: no method matching show_datatype(::IOContext{REPL.Terminals.TTYTerminal}, ::Type{Enum})
Closest candidates are:
  show_datatype(::IO, ::DataType) at show.jl:596
Stacktrace:
 [1] show(::IOContext{REPL.Terminals.TTYTerminal}, ::MIME{Symbol("text/plain")}, ::Type{Enum}) at ./Enums.jl:52
 [2] display(::REPL.REPLDisplay{REPL.LineEditREPL}, ::MIME{Symbol("text/plain")}, ::Type{T} where T) at /home/marc/.julia/packages/OhMyREPL/7qVFi/src/output_prompt_overwrite.jl:6
 [3] display(::REPL.REPLDisplay, ::Any) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/REPL/src/REPL.jl:218
 [4] display(::Any) at ./multimedia.jl:328
 [5] #invokelatest#1 at ./essentials.jl:710 [inlined]
 [6] invokelatest at ./essentials.jl:709 [inlined]
 [7] print_response(::IO, ::Any, ::Bool, ::Bool, ::Any) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/REPL/src/REPL.jl:238
 [8] print_response(::REPL.AbstractREPL, ::Any, ::Bool, ::Bool) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/REPL/src/REPL.jl:223
 [9] (::REPL.var"#do_respond#54"{Bool,Bool,REPL.var"#64#73"{REPL.LineEditREPL,REPL.REPLHistoryProvider},REPL.LineEditREPL,REPL.LineEdit.Prompt})(::Any, ::Any, ::Any) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/REPL/src/REPL.jl:822
 [10] #invokelatest#1 at ./essentials.jl:710 [inlined]
 [11] invokelatest at ./essentials.jl:709 [inlined]
 [12] run_interface(::REPL.Terminals.TextTerminal, ::REPL.LineEdit.ModalInterface, ::REPL.LineEdit.MIState) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/REPL/src/LineEdit.jl:2355
 [13] run_frontend(::REPL.LineEditREPL, ::REPL.REPLBackendRef) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/REPL/src/REPL.jl:1143
 [14] (::REPL.var"#38#42"{REPL.LineEditREPL,REPL.REPLBackendRef})() at ./task.jl:356

julia> versioninfo()
Julia Version 1.5.0
Commit 96786e22cc (2020-08-01 23:44 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)
```

### PR DESCRIPTION
**Pull request by KristofferC**
Backported PRs:
- [x] #36810 - Define `OrderStyle` for `Union{}`
- [x] #37130 - Fix string interpolation into Markdown.Table and Markdown.List
- [x] #37178 - fix #37126, spurious soft scope warning on hidden name
- [x] #37181 - fix #37134, error with macro returning `x::T -> y`
- [x] #37204 - fix #37199, indexing for 1-d views with offset ranges
- [x] #37315 - reset ACTIVE_PROJECT when building docs
- [x] #37357 - Mark that ismutable requires julia 1.5
- [x] #37385 - Make sure to build clangsa before running clangsa self tests
- [x] #37296 - Allow packages to force depwarns
- [x] #37404 - fix #37401, displaying `Enum` in the repl
- [x] #37149 - fix ccall of strlen
- [x] #37114 - Improve optimizations on NULL pointer check
- [x] #37365 - inference: fix precision regression
- [x] #37308 - fix #37282, printing juxtaposition of 0 confusion with hex,bin,oct literals
- [x] #37386 - fix a possible segfault during static compilation
- [x] #37394 - Build `julia_version.h` before running `clangsa`
- [x] #37285 - Only emit incoming values for the phi when needed
- [x] #37319 - codegen: phic/upsilon nodes can be set to undefined
- [x] #37388 - Fix use of CreateAlignedStore
- [x] #37443 - Fix `getfield_tfunc` for constant `TypeName` 
- [x] #37444 - Specialize _memory_offset on number of arguments.
- [x] #37261 - Fix type-equality involving `Type{Union{}}` 
- [x] #37207 - Prune egal types when forming a `Union` 
- [x] #37472  - Fix memory leak in umfpack_numeric! 
- [x] #37534 - fix "invalid assignment location" macro hygiene regression
- [x] #37511 - Fix getfield of potentially undef inline immutable field
- [x] bump Pkg to 1.5.2 

Contains multiple commits, manual intervention needed:
- [ ] #37243 - code coverage fixes

Non-merged PRs with backport label:
- [ ] #36821 - Fix function call to rec_backtrace_ctx when using USE_SYSTEM_LIBUNWIND=1

### CODE REVIEW

**Comment by yuyichao** [comment]
Seems that the back point of https://github.com/JuliaLang/julia/pull/37149 included more than just the bug fix. Only e change to gcutils.jl and secretbuffer.jl should be backported since those fixed the wrong return type from strlen call...

**Comment by KristofferC** [comment]
Thanks, I missed that comment, will change it.

File: test/compiler/inference.jl [hunk 1]
```diff
@@ -2617,3 +2625,25 @@ f36531(args...) = tuple((args...)...)
...
+# Check constant propagation for field of constant `TypeName`
+# works for both valid and invalid field names. (Ref #37443)
+getfield_const_typename_good1() = getfield(const_int32_typename, 1)
+getfield_const_typename_good2() = getfield(const_int32_typename, :name)
+getfield_const_typename_bad1() = getfield(const_int32_typename, 0x1)
+@eval getfield_const_typename_bad2() = getfield(const_int32_typename, $(()))
+for goodf in [getfield_const_typename_good1, getfield_const_typename_good2]
+    local goodf
+    local code = code_typed(goodf, Tuple{})[1].first.code
+    @test code[1] === Core.ReturnNode(QuoteNode(:Int32))
```
Comment (martinholters) [inline test/compiler/inference.jl:2640]:
```suggestion
    @test code[1] === Core.Compiler.ReturnNode(QuoteNode(:Int32))
```

File: test/compiler/inference.jl [hunk 2]
```diff
@@ -2617,3 +2625,25 @@ f36531(args...) = tuple((args...)...)
...
+    @test goodf() === :Int32
+end
+for badf in [getfield_const_typename_bad1, getfield_const_typename_bad2]
+    local badf
+    local code = code_typed(badf, Tuple{})[1].first.code
+    @test Meta.isexpr(code[1], :call)
+    @test code[end] === Core.ReturnNode()
```
Comment (martinholters) [inline test/compiler/inference.jl:2647]:
```suggestion
    @test code[end] === Core.Compiler.ReturnNode()
```

**Review by martinholters — COMMENTED** [review]
The tests from #37443 need a minor adjustment.

**Comment by JeffBezanson** [comment]
There is a test failure here from 38c3e741af5e18d41f0420f5fe56a8a54b1ee248. It looks like some part of dcc0696971a100dbc63f90b8789632631dd1bfef is also needed to make that test work. That whole commit seems likely safe to backport though.

**Comment by JeffBezanson** [comment]
This is the needed line:
```
--- a/base/compiler/inferencestate.jl
+++ b/base/compiler/inferencestate.jl
@@ -205,7 +205,7 @@ update_valid_age!(edge::InferenceState, sv::InferenceState) = update_valid_age!(
 function record_ssa_assign(ssa_id::Int, @nospecialize(new), frame::InferenceState)
     old = frame.src.ssavaluetypes[ssa_id]
     if old === NOT_FOUND || !(new ⊑ old)
-        frame.src.ssavaluetypes[ssa_id] = tmerge(old, new)
+        frame.src.ssavaluetypes[ssa_id] = old === NOT_FOUND ? new : tmerge(old, new)
         W = frame.ip
         s = frame.stmt_types
```

**Comment by martinholters** [comment]
Ref. https://github.com/JuliaLang/julia/pull/37365#issuecomment-688721730 :wink:

**Comment by KristofferC** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":release-1.5")`

**Comment by KristofferC** [comment]
@nanosoldier `runtests(ALL, vs = ":release-1.5")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/37210#issuecomment-689410445) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/40559b5cbc31f3c53c2e6af7cf9900fe70c42d06/be8475f_vs_c4acbf9/report.md). cc @ararslan

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/37210#issuecomment-689410535) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/801671c4f9782fadd4f9b6718570c91a1414fd0e/pkgeval/by_hash/21d877d_vs_c4acbf9/report.md). cc @maleadt

**Comment by KristofferC** [comment]
https://github.com/JuliaCI/NanosoldierReports/blob/801671c4f9782fadd4f9b6718570c91a1414fd0e/pkgeval/by_hash/21d877d_vs_c4acbf9/logs/StatProfilerHTML/1.5.2-pre-be8475f41a.log

```
ERROR: LoadError: LoadError: syntax: invalid assignment location "#40#content..." around /home/pkgeval/.julia/packages/HAML/hMyIa/src/Templates.jl:49
Stacktrace:
 [1] top-level scope at none:1
 [2] eval at ./boot.jl:331 [inlined]
 [3] eval at ./Base.jl:39 [inlined]
 [4] _includehaml(::Module, ::Symbol, ::String, ::String) at /home/pkgeval/.julia/packages/HAML/hMyIa/src/Templates.jl:98
 [5] includehaml at /home/pkgeval/.julia/packages/HAML/hMyIa/src/Templates.jl:68 [inlined] (repeats 2 times)
 [6] #3 at /home/pkgeval/.julia/packages/HAML/hMyIa/src/Templates.jl:71 [inlined]
 [7] foreach at ./abstractarray.jl:2009 [inlined]
 [8] includehaml(::Module, ::Pair{Symbol,String}, ::Pair{Symbol,String}, ::Pair{Symbol,String}, ::Pair{Symbol,String}, ::Vararg{Pair{Symbol,String},N} where N) at /home/pkgeval/.julia/packages/HAML/hMyIa/src/Templates.jl:70
 [9] top-level scope at /home/pkgeval/.julia/packages/StatProfilerHTML/BuyY0/src/HTML.jl:12
 [10] include(::Function, ::Module, ::String) at ./Base.jl:380
 [11] include at ./Base.jl:368 [inlined]
 [12] include(::String) at /home/pkgeval/.julia/packages/StatProfilerHTML/BuyY0/src/StatProfilerHTML.jl:1
 [13] top-level scope at /home/pkgeval/.julia/packages/StatProfilerHTML/BuyY0/src/StatProfilerHTML.jl:6
 [14] include(::Function, ::Module, ::String) at ./Base.jl:380
```

looks a bit strange. @JeffBezanson?

**Comment by fredrikekre** [comment]
Backport of https://github.com/JuliaLang/julia/pull/37315 should be `HOME_PROJECT` on 1.5 (https://github.com/JuliaLang/julia/pull/37315#pullrequestreview-479690267)

**Comment by KristofferC** [comment]
This should be good to go once CI finishes.

**Pull request merged**